### PR TITLE
Some preparatory commits for the lazy data loading feature of Information Schema Connector

### PR DESCRIPTION
**Pull request by linxingyuan1102**
cc: @kokosing

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java [hunk 1]
```diff
@@ -71,17 +71,8 @@
...
-    public static final SchemaTableName TABLE_VIEWS = new SchemaTableName(INFORMATION_SCHEMA, "views");
-    public static final SchemaTableName TABLE_SCHEMATA = new SchemaTableName(INFORMATION_SCHEMA, "schemata");
-    public static final SchemaTableName TABLE_TABLE_PRIVILEGES = new SchemaTableName(INFORMATION_SCHEMA, "table_privileges");
-    public static final SchemaTableName TABLE_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "roles");
-    public static final SchemaTableName TABLE_APPLICABLE_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "applicable_roles");
-    public static final SchemaTableName TABLE_ENABLED_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "enabled_roles");
-
     public static final Map<SchemaTableName, ConnectorTableMetadata> TABLES = schemaMetadataBuilder()
-            .table(tableMetadataBuilder(TABLE_COLUMNS)
+            .table(tableMetadataBuilder(InformationSchemaTableKind.COLUMNS.getSchemaTableName())
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java]: static imports

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java [hunk 2]
```diff
@@ -71,17 +71,8 @@
...
-    public static final SchemaTableName TABLE_COLUMNS = new SchemaTableName(INFORMATION_SCHEMA, "columns");
-    public static final SchemaTableName TABLE_TABLES = new SchemaTableName(INFORMATION_SCHEMA, "tables");
-    public static final SchemaTableName TABLE_VIEWS = new SchemaTableName(INFORMATION_SCHEMA, "views");
-    public static final SchemaTableName TABLE_SCHEMATA = new SchemaTableName(INFORMATION_SCHEMA, "schemata");
-    public static final SchemaTableName TABLE_TABLE_PRIVILEGES = new SchemaTableName(INFORMATION_SCHEMA, "table_privileges");
-    public static final SchemaTableName TABLE_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "roles");
-    public static final SchemaTableName TABLE_APPLICABLE_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "applicable_roles");
-    public static final SchemaTableName TABLE_ENABLED_ROLES = new SchemaTableName(INFORMATION_SCHEMA, "enabled_roles");
-
     public static final Map<SchemaTableName, ConnectorTableMetadata> TABLES = schemaMetadataBuilder()
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java]: Maybe this map should also be extracted. You could move `ConnectorTableMetadata` field to `InformationSchemaTableKind`. What do you think?

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java [hunk 3]
```diff
@@ -303,9 +294,12 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         return prefixes;
     }
 
-    private boolean isTablesEnumeratingTable(SchemaTableName schemaTableName)
+    private boolean isTablesEnumeratingTable(InformationSchemaTableKind tableKind)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java]: that could be also a property in `InformationSchemaTableKind`

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTableKind.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import io.prestosql.spi.connector.SchemaTableName;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.prestosql.connector.informationschema.InformationSchemaMetadata.INFORMATION_SCHEMA;
+import static java.util.Locale.ENGLISH;
+
+public enum InformationSchemaTableKind
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTableKind.java]: s/InformationSchemaTableKind/InformationSchemaTable ?

File: presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java [hunk 5]
```diff
@@ -205,7 +200,7 @@ protected Node visitShowTables(ShowTables showTables, Void context)
 
             return simpleQuery(
                     selectList(aliasedName("table_name", "Table")),
-                    from(schema.getCatalogName(), TABLE_TABLES),
+                    from(schema.getCatalogName(), InformationSchemaTableKind.TABLES),
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java]: static imports

**Comment by linxingyuan1102** [comment]
@kokosing Thanks for the review. I'll update this PR after #1306 gets merged

File: presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                public Iterable<ConnectorFactory> getConnectorFactories()
+                {
+                    List<SchemaTableName> tablesTestSchema1 = IntStream.range(0, 100000)
+                            .mapToObj(i -> new SchemaTableName("test_schema1", "test_table" + i))
+                            .collect(toImmutableList());
+                    List<SchemaTableName> tablesTestSchema2 = IntStream.range(0, 200000)
+                            .mapToObj(i -> new SchemaTableName("test_schema2", "test_table" + i))
+                            .collect(toImmutableList());
+                    List<SchemaTableName> tables = Stream.concat(tablesTestSchema1.stream(), tablesTestSchema2.stream()).collect(toImmutableList());
+                    MockConnectorFactory.Builder builder = MockConnectorFactory.builder();
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java]: inline builder

File: presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                    MockConnectorFactory mockConnectorFactory = builder.withListSchemaNames(connectorSession -> ImmutableList.of("test_schema1", "test_schema2"))
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java]: %s/connectorSession/session

File: presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        assertQuery("SELECT count(*) from test_catalog.information_schema.tables WHERE table_schema = 'test_schema1'", "VALUES 100000");
+        assertQuery("SELECT count(*) from test_catalog.information_schema.tables WHERE table_name = 'test_table1'", "VALUES 2");
+    }
+
+    private static DistributedQueryRunner createQueryRunner()
+            throws Exception
+    {
+        Session session = testSessionBuilder().build();
+        DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(session)
+                .setNodeCount(4)
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java]: why do you need 4 nodes?

Comment (linxingyuan1102) [reply presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java]: The default value is 4. I'll remove this line.

Comment (kokosing) [reply presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java]: Let me take care of this with: https://github.com/prestosql/presto/pull/1334

Comment (linxingyuan1102) [reply presto-tests/src/test/java/io/prestosql/tests/TestInformationSchemaConnector.java]: thanks

**Comment by kokosing** [comment]
> I'll update this PR after #1306 gets merged

#1306 is merged, can you please rebase and address remaining comments?

**Comment by linxingyuan1102** [comment]
> > I'll update this PR after #1306 gets merged
> 
> #1306 is merged, can you please rebase and address remaining comments?

@kokosing 
Sure, I've rebased it against the latest master branch. I addressed most of the comments except the one suggesting making `isTablesEnumeratingTable` a property in `InformationSchemaTable` (previously named `InformationSchemaTableKind`). I think it kind of makes the code less cleaner so am not sure if it is a good idea (especially we also have `isColumnsEnumeratingTable` now).

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java [hunk 9]
```diff
@@ -172,37 +97,36 @@ private InformationSchemaTableHandle checkTableHandle(ConnectorTableHandle table
...
-        if (!schemaName.isPresent()) {
-            return ImmutableList.copyOf(TABLES.keySet());
+        if (schemaName.isPresent() && !schemaName.get().equals(INFORMATION_SCHEMA)) {
+            return ImmutableList.of();
         }
-
-        return TABLES.keySet().stream()
-                .filter(compose(schemaName.get()::equals, SchemaTableName::getSchemaName))
-                .collect(toImmutableList());
+        return Arrays.stream(InformationSchemaTable.values()).map(InformationSchemaTable::getSchemaTableName).collect(toImmutableList());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java]:
```
        return Arrays.stream(InformationSchemaTable.values())
            .map(InformationSchemaTable::getSchemaTableName)
            .collect(toImmutableList());
````

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java [hunk 10]
```diff
@@ -228,9 +152,9 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     {
         requireNonNull(prefix, "prefix is null");
         ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> builder = ImmutableMap.builder();
-        for (Entry<SchemaTableName, ConnectorTableMetadata> entry : TABLES.entrySet()) {
-            if (prefix.matches(entry.getKey())) {
-                builder.put(entry.getKey(), entry.getValue().getColumns());
+        for (InformationSchemaTable table : InformationSchemaTable.values()) {
+            if (prefix.matches(table.getSchemaTableName())) {
+                builder.put(table.getSchemaTableName(), table.getTableMetadata().getColumns());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java]:
```
        return Arrays.stream(InformationSchemaTable.values())
            .filter(prefix::matches)
            .collect(toImmutableMap(...));
```

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java [hunk 11]
```diff
@@ -172,37 +97,36 @@ private InformationSchemaTableHandle checkTableHandle(ConnectorTableHandle table
     @Override
     public ConnectorTableHandle getTableHandle(ConnectorSession connectorSession, SchemaTableName tableName)
     {
-        if (!TABLES.containsKey(tableName)) {
+        try {
+            InformationSchemaTable informationSchemaTable = InformationSchemaTable.of(tableName);
+            return new InformationSchemaTableHandle(catalogName, informationSchemaTable, defaultPrefixes(), OptionalLong.empty());
+        }
+        catch (IllegalArgumentException e) {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java]: I would prefer use `Optional` instead throwing an exception

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+                    .column("is_nullable", createUnboundedVarcharType())
+                    .column("data_type", createUnboundedVarcharType())
+                    .column("comment", createUnboundedVarcharType())
+                    .column("extra_info", createUnboundedVarcharType())
+                    .hiddenColumn("column_comment", createUnboundedVarcharType()) // MySQL compatible
+                    .build();
+            return;
+        }
+        if (tableName.equals("tables")) {
+            tableMetadata = tableMetadataBuilder(schemaTableName)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java]: maybe tableMetadata could be a parameter to constructor?

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    public static InformationSchemaTable of(SchemaTableName schemaTableName)
+    {
+        checkArgument(schemaTableName.getSchemaName().equals(INFORMATION_SCHEMA));
+        try {
+            return InformationSchemaTable.valueOf(schemaTableName.getTableName().toUpperCase(ENGLISH));
+        }
+        catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException("table does not exist: " + schemaTableName);
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java]: `return Optional?`

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+    @JsonCreator
+    public InformationSchemaTable fromString(String value)
+    {
+        return InformationSchemaTable.valueOf(value);
+    }
+
+    @JsonValue
+    @Override
+    public String toString()
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java]: let's not use `toString` for this. `toStringValue` and `fromStringValue`?

File: presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java (hunk 12 above)
Comment (linxingyuan1102) [reply presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaTable.java]: I thought about this but found it clumsy to do so. For example, `tableMetadataBuilder` takes the `schemaTableName` as an argument. It looks cleaner to have it initialized firstly. So should we just leave the building inside the constructor?

**Comment by linxingyuan1102** [comment]
Thanks. I've updated the PR

**Comment by kokosing** [comment]
There are test failures that seems to be legit:
```
2019-09-05 11:24:29 INFO: [375 of 435] sql_tests.testcases.system.selectInformationSchemaColumns (Groups: system)
14 - expected: system|information_schema|enabled_roles|role_name|varchar|YES|null|null|
14 - actual:   system|information_schema|enable_roles|role_name|varchar|YES|null|null|
```

It looks like accidentally you changed the table name.

**Comment by linxingyuan1102** [comment]
That's true. I updated the PR and now the tests pass.

**Comment by kokosing** [comment]
Merged, thanks!

**Pull request merged**